src/IRTS/JavaScript/AST.hs

```diff
@@ -1,409 +1,208 @@
-
-compileJS' indent (JSWhile cond body) =
-     "while (" `T.append` compileJS' indent cond `T.append` ") {\n"
-  `T.append` compileJS' (indent + 2) body
-  `T.append` "\n" `T.append` T.replicate indent " " `T.append` "}"
-
-compileJS' indent (JSWord word)
-  | JSWord8  b <- word =
-      "new Uint8Array([" `T.append` T.pack (show b) `T.append` "])"
-  | JSWord16 b <- word =
-      "new Uint16Array([" `T.append` T.pack (show b) `T.append` "])"
-  | JSWord32 b <- word =
-      "new Uint32Array([" `T.append` T.pack (show b) `T.append` "])"
-  | JSWord64 b <- word =
-      "i$bigInt(\"" `T.append` T.pack (show b) `T.append` "\")"
-
-jsInstanceOf :: JS -> String -> JS
-jsInstanceOf obj cls = JSBinOp "instanceof" obj (JSIdent cls)
-
-jsOr :: JS -> JS -> JS
-jsOr lhs rhs = JSBinOp "||" lhs rhs
-
-jsAnd :: JS -> JS -> JS
-jsAnd lhs rhs = JSBinOp "&&" lhs rhs
-
-jsMeth :: JS -> String -> [JS] -> JS
-jsMeth obj meth args = JSApp (JSProj obj meth) args
-
-jsCall :: String -> [JS] -> JS
-jsCall fun args = JSApp (JSIdent fun) args
-
-jsTypeOf :: JS -> JS
-jsTypeOf js = JSPreOp "typeof " js
-
-jsEq :: JS -> JS -> JS
-jsEq lhs@(JSNum (JSInteger _)) rhs = JSApp (JSProj lhs "equals") [rhs]
-jsEq lhs rhs@(JSNum (JSInteger _)) = JSApp (JSProj lhs "equals") [rhs]
-jsEq lhs rhs = JSBinOp "==" lhs rhs
-
-jsNotEq :: JS -> JS -> JS
-jsNotEq lhs rhs = JSBinOp "!=" lhs rhs
-
-jsIsNumber :: JS -> JS
-jsIsNumber js = (jsTypeOf js) `jsEq` (JSString "number")
-
-jsIsNull :: JS -> JS
-jsIsNull js = JSBinOp "==" js JSNull
-
-jsBigInt :: JS -> JS
-jsBigInt (JSString "0") = JSNum (JSInteger JSBigZero)
-jsBigInt (JSString "1") = JSNum (JSInteger JSBigOne)
-jsBigInt js             = JSNum $ JSInteger $ JSBigIntExpr js
-
-jsUnPackBits :: JS -> JS
-jsUnPackBits js = JSIndex js $ JSNum (JSInt 0)
-
-jsPackUBits8 :: JS -> JS
-jsPackUBits8 js = JSNew "Uint8Array" [JSArray [js]]
-
-jsPackUBits16 :: JS -> JS
-jsPackUBits16 js = JSNew "Uint16Array" [JSArray [js]]
-
-jsPackUBits32 :: JS -> JS
-jsPackUBits32 js = JSNew "Uint32Array" [JSArray [js]]
-
-jsPackSBits8 :: JS -> JS
-jsPackSBits8 js = JSNew "Int8Array" [JSArray [js]]
-
-jsPackSBits16 :: JS -> JS
-jsPackSBits16 js = JSNew "Int16Array" [JSArray [js]]
-
-jsPackSBits32 :: JS -> JS
-jsPackSBits32 js = JSNew "Int32Array" [JSArray [js]]
+jsAst2Text (JsComment c)=
+  T.concat ["/*",c,"*/"]
+jsAst2Text (JsLazy e) =
+  T.concat [ "{js_idris_lazy_calc:function(){"
+           , indent $ jsAst2Text e
+           , "}}"
+           ]
+jsAst2Text (JsForce e) =
+  T.concat ["js_idris_force(", jsAst2Text e, ")"]
+
+
+
+case2Text :: (JsAST, JsAST) -> Text
+case2Text (x,y) =
+  T.concat [ "case ", jsAst2Text x, ":\n"
+           , indent $ T.concat [ jsAst2Text y, ";\nbreak;\n"]
+           ]
+
+throw2 = T.concat [ "var js_idris_throw2 = function (x){\n"
+                  , " throw x;\n"
+                  , "}\n\n"
+                  ]
+
+force  = T.concat [ "var js_idris_force = function (x){\n"
```

Comment:
Not a good one. When I wrote it this was a separated backend and because it was so small code this seamed the easier way. I am happy to change it. Right now it is not possible, but probably today.